librz/analysis/dwarf_process.c

```diff
@@ -1943,7 +1943,6 @@ static bool dwarf_integrate_function(void *user, const ut64 k, const void *value
 	if (callable) {
 		char *sig = rz_type_callable_as_string(analysis->typedb, callable);
 		rz_meta_set_string(analysis, RZ_META_TYPE_COMMENT, dw_fn->low_pc, sig);
-		free(sig);
```

Comment:
I don't see why it would cause double free.

Reply:
i checked the code. i agree with billow.

Reply:
Initially I thought that too, but LLDB doesn't lie, without this patch this is what I get:
```
[0x80080020]> q
rizin(68472,0x1dff6d300) malloc: *** error for object 0x189: pointer being freed was not allocated
rizin(68472,0x1dff6d300) malloc: *** set a breakpoint in malloc_error_break to debug
Process 68472 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x0000000188c6f11c libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`:
->  0x188c6f11c <+8>:  b.lo   0x188c6f13c               ; <+40>
    0x188c6f120 <+12>: pacibsp
    0x188c6f124 <+16>: stp    x29, x30, [sp, #-0x10]!
    0x188c6f128 <+20>: mov    x29, sp
Target 0: (rizin) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x0000000188c6f11c libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x0000000188ca6cc0 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x0000000188bb6a50 libsystem_c.dylib`abort + 180
    frame #3: 0x0000000188acdb08 libsystem_malloc.dylib`malloc_vreport + 908
    frame #4: 0x0000000188ad13f4 libsystem_malloc.dylib`malloc_report + 64
    frame #5: 0x0000000188ae5ebc libsystem_malloc.dylib`find_zone_and_free + 308
    frame #6: 0x0000000100b56530 librz_analysis.0.7.dylib`rz_meta_item_free [inlined] rz_meta_item_fini(item=0x000060003628b140) at analysis.c:48:2 [opt]
    frame #7: 0x0000000100b56528 librz_analysis.0.7.dylib`rz_meta_item_free(_item=0x000060003628b140) at analysis.c:54:3 [opt]
    frame #8: 0x00000001005a62e8 librz_util.0.7.dylib`interval_node_free(node=0x0000600003807d80, user=<unavailable>) at intervaltree.c:109:3 [opt]
    frame #9: 0x00000001005b602c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003945a80, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:286:2 [opt]
    frame #10: 0x00000001005b601c librz_util.0.7.dylib`rz_rbtree_free(x=0x000060000381ce00, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:285:2 [opt]
    frame #11: 0x00000001005b601c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003813800, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:285:2 [opt]
    frame #12: 0x00000001005b601c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003a89b80, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:285:2 [opt]
    frame #13: 0x00000001005b602c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003ab0700, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:286:2 [opt]
    frame #14: 0x00000001005b601c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003af6940, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:285:2 [opt]
    frame #15: 0x00000001005b602c librz_util.0.7.dylib`rz_rbtree_free(x=0x000060000391c780, freefn=(librz_util.0.7.dylib`interval_node_free at intervaltree.c:106), user=0x0000000100b56514) at rbtree.c:286:2 [opt]
    frame #16: 0x00000001005b602c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003aba380, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:286:2 [opt]
    frame #17: 0x00000001005b601c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003abb600, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:285:2 [opt]
    frame #18: 0x00000001005b601c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003ab7080, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:285:2 [opt]
    frame #19: 0x00000001005b602c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003946580, freefn=(librz_util.0.7.dylib`interval_node_free at intervaltree.c:106), user=0x0000000100b56514) at rbtree.c:286:2 [opt]
    frame #20: 0x00000001005b602c librz_util.0.7.dylib`rz_rbtree_free(x=0x0000600003fe4dc0, freefn=<unavailable>, user=0x0000000100b56514) at rbtree.c:286:2 [opt]
    frame #21: 0x00000001005a62c4 librz_util.0.7.dylib`rz_interval_tree_fini(tree=<unavailable>) at intervaltree.c:118:2 [opt] [artificial]
    frame #22: 0x0000000100b568b0 librz_analysis.0.7.dylib`rz_analysis_free(a=0x0000000132009800) at analysis.c:149:2 [opt]
    frame #23: 0x0000000102423710 librz_core.0.7.dylib`rz_core_fini(c=0x0000000132808200) at core.c:2594:2 [opt]
    frame #24: 0x0000000102423828 librz_core.0.7.dylib`rz_core_free(c=0x0000000132808200) at core.c:2623:2 [opt]
    frame #25: 0x000000010049eeec librz_main.0.7.dylib`rz_main_rizin(argc=<unavailable>, argv=0x000000016fdfea48) at rizin.c:1530:2 [opt]
    frame #26: 0x000000018892d058 dyld`start + 2224
(lldb) frame select 7
warning: librz_analysis.0.7.dylib was compiled with optimization - stepping may behave oddly; variables may not be available.
frame #7: 0x0000000100b56528 librz_analysis.0.7.dylib`rz_meta_item_free(_item=0x000060003628b140) at analysis.c:54:3 [opt]
   51   static void rz_meta_item_free(void *_item) {
   52           if (_item) {
   53                   RzAnalysisMetaItem *item = _item;
-> 54                   rz_meta_item_fini(item);
   55                   free(item);
   56           }
   57   }
(lldb) p item->str
(char *) $0 = 0x0000000000000189 ""
(lldb) p item->size
(size_t) $1 = 0
(lldb)
```

Reply:
And I tracked where it was allocated - it was allocated in that dwarf integration function.

Reply:
I agree it does not seem right at first sight... maybe there's something else going on and the real problem is elsewhere.

Reply:
it means that `rz_type_callable_as_string` is not returning a copy

Reply:
Now this change is unnecessary